api/python/quilt3/data_transfer.py

```diff
@@ -867,20 +942,29 @@ def _calculate_sha256_internal(src_list, sizes, results):
         for size, result in zip(sizes, results)
         if result is None or isinstance(result, Exception)
     )
-    lock = Lock()
-
-    with tqdm(desc="Hashing", total=total_size, unit='B', unit_scale=True, disable=DISABLE_TQDM) as progress:
-        def _process_url(src, size):
-            hash_obj = hashlib.sha256()
-            if src.is_local():
+    s3_client_provider = S3ClientProvider()
+    # This controls how many parts can be stored in the memory.
+    # This includes the ones that are being downloaded or hashed.
+    s3_max_pending_parts = s3_transfer_config.max_request_concurrency * 4
+
+    async def main():
+        loop = asyncio._get_running_loop()
+
+        s3_pending_parts_semaphore = asyncio.BoundedSemaphore(s3_max_pending_parts)
+        with tqdm(desc="Hashing", total=total_size, unit='B', unit_scale=True, disable=DISABLE_TQDM) as progress, \
```

Comment:
why not `unit=MB`? that's what we do today

Reply:
I think tqdm adds prefixes automatically, so it should be `M`.

Reply:
> unit_scale: bool or int or float, optional
If 1 or True, the number of iterations will be printed with an appropriate SI metric prefix (k = 10^3, M = 10^6, etc.) [default: False]. If any other non-zero number, will scale total and n.